### PR TITLE
Improved caching of state summaries

### LINKED ISSUE #11384: ERROR main: could not get ancestor state: signed beacon block can't be nil
# 🐞 Bug Report

### Description

Node Can't Be Started.

### Has this worked before in a previous version?

Yes

## 🔬 Minimal Reproduction

````zsh
./prysm.sh beacon-chain --datadir=/mnt/data/Eth2 --execution-endpoint=http://<SNIP>:8551 --mainnet --jwt-secret=/mnt/data/jwt.hex --suggested-fee-recipient=<SNIP> --monitoring-host=0.0.0.0 --monitoring-port=6070
````

## 🔥 Error

````zsh
Latest Prysm version is v3.0.0.
Beacon chain is up to date.
Verifying binary integrity.
beacon-chain-v3.0.0-linux-amd64: OK
gpg: Signature made Mon 22 Aug 2022 08:50:44 PM UTC
gpg:                using RSA key 0AE0051D647BA3C1A917AF4072E33E4DF1A5036E
gpg: Good signature from "Preston Van Loon <preston@prysmaticlabs.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 0AE0 051D 647B A3C1 A917  AF40 72E3 3E4D F1A5 036E
Verified /mnt/data/dist/beacon-chain-v3.0.0-linux-amd64 has been signed by Prysmatic Labs.
Starting Prysm beacon-chain --datadir=/mnt/data/Eth2 --execution-endpoint=http://<SNIP>:8551 --mainnet --jwt-secret=/mnt/data/jwt.hex --suggested-fee-recipient=<SNIP> --monitoring-host=0.0.0.0 --monitoring-port=6070
[2022-09-01 12:53:42]  WARN flags: Running on Ethereum Mainnet
[2022-09-01 12:53:42]  INFO node: Checking DB database-path=/mnt/data/Eth2/beaconchaindata
[2022-09-01 12:53:42]  INFO db: Opening Bolt DB at /mnt/data/Eth2/beaconchaindata/beaconchain.db
[2022-09-01 12:53:43]  INFO node: Deposit contract: 0x00000000219ab540356cbb839cbe05303d7705fa
[2022-09-01 12:53:43] ERROR main: could not get ancestor state: signed beacon block can't be nil
````

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Ubuntu Server
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
v3.0.0
  </code>
</pre>

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What does this PR do? Why is it needed?**

The DB must ensure that blocks referenced by state summaries exist in the database. 

**Which issues(s) does this PR fix?**

Fixes #11384

**Other notes for review**

Do not merge yet!

Key changes are:

- Addition of saving of state summaries with a block cache accessor. This is something that initial sync would use to attempt save state summaries with blocks that the database doesn't know about yet. `beacon-chain/db/kv/state_summary.go`
- Initial sync routine using this new method `beacon-chain/blockchain/process_block.go`
- Enforcement that flushing the state summary buffer has access to blocks in the database. `beacon-chain/db/kv/state_summary.go`

### CODE REVIEW

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 1]
```diff
@@ -75,6 +75,12 @@ func (s *Service) getInitSyncBlocks() []interfaces.SignedBeaconBlock {
 	return blks
 }
 
+func (s *Service) getInitSyncBlockFromCache(r [32]byte) interfaces.SignedBeaconBlock {
+	s.initSyncBlocksLock.RLock()
+	defer s.initSyncBlocksLock.RUnlock()
+	return s.initSyncBlocks[r]
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/init_sync_process_block.go:81]: No, nil is a cache miss

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -408,26 +412,39 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []interfaces.SignedBeac
...
+	for _, c := range justified {
+		if err := s.cfg.BeaconDB.SaveJustifiedCheckpoint(ctx, c); err != nil {
+			tracing.AnnotateError(span, err)
+			return err
+		}
+	}
+	for _, c := range finalized {
+		if err := s.updateFinalized(ctx, c); err != nil {
+			tracing.AnnotateError(span, err)
+			return err
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/process_block.go:444]: Yes. I wanted to save all of the state summaries at once but you can't update these things until after the state summaries have been saved. It is a bit of a bad design for sure.

File: beacon-chain/blockchain/init_sync_process_block.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/init_sync_process_block.go:81]: Actually, I see that the other getter method returns an error. Perhaps that is  for safety reasons. I will amend my signature to match the other getter

File: beacon-chain/db/kv/state_summary.go [hunk 3]
```diff
@@ -37,6 +49,25 @@ func (s *Store) SaveStateSummaries(ctx context.Context, summaries []*ethpb.State
 	return nil
 }
 
+func (s *Store) ensureBlocksSaved(ctx context.Context, blockCache func([32]byte) interfaces.SignedBeaconBlock) error {
+	if blockCache == nil {
+		return nil
+	}
+	blocks := make([]interfaces.SignedBeaconBlock, 0)
+	summaries := s.stateSummaryCache.getAll()
```
Comment (terencechain) [inline beacon-chain/db/kv/state_summary.go:57]: wdyt?

```suggestion
	summaries := s.stateSummaryCache.getAll()
	blocks := make([]interfaces.SignedBeaconBlock, 0, len(summaries))
```

File: beacon-chain/db/kv/state_summary.go [hunk 4]
```diff
@@ -37,6 +49,25 @@ func (s *Store) SaveStateSummaries(ctx context.Context, summaries []*ethpb.State
...
+	for _, ss := range summaries {
+		if s.HasBlock(ctx, bytesutil.ToBytes32(ss.Root)) {
+			log.WithField("blockRoot", fmt.Sprintf("%#x", ss.Root)).Debug("Block already saved")
```
Comment (terencechain) [inline beacon-chain/db/kv/state_summary.go:60]: This log may be noisy even under debug. I'm not sure though...

File: beacon-chain/blockchain/process_block.go [hunk 5]
```diff
@@ -408,26 +412,39 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []interfaces.SignedBeac
...
+			finalized = append(finalized, fCheckpoints[i])
+		}
+	}
+
+	if err := s.cfg.BeaconDB.SaveStateSummariesWithPendingBlocks(ctx, summaries, s.getInitSyncBlockFromCache); err != nil {
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	// TODO: Batch these transactions?
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:434]: TODO with an issue number please

File: beacon-chain/db/kv/blocks.go [hunk 6]
```diff
@@ -296,32 +296,40 @@ func (s *Store) SaveBlocks(ctx context.Context, blks []interfaces.SignedBeaconBl
...
 				return err
 			}
 		}
 		return nil
-	})
+	}); err != nil {
+		return err
+	}
+
+	// Populate cache only after successful db update.
```
Comment (terencechain) [inline beacon-chain/db/kv/blocks.go:318]: Nice!

File: beacon-chain/blockchain/process_block.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/blockchain/process_block.go:444]: Not clear to me whether that is saving state summaries at once for performance or the bug fix. If it's for performance, what is the rough gain? 
A more conservative approach here would be  `SaveStateSummaryWithPendingBlock` which is just the status quo

[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/process_block.go:444]: Given the way that state storage summary buffer is implemented, these changes are probably a performance regression. I'll change it back so its a minimal change here. Thanks

File: beacon-chain/db/kv/state_summary.go (hunk 4 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/state_summary.go:60]: Yeah, i'll delete it

File: beacon-chain/blockchain/process_block.go (hunk 5 above)
Comment (prestonvanloon) [reply beacon-chain/blockchain/process_block.go:434]: Will revert

File: beacon-chain/db/kv/blocks.go (hunk 6 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/blocks.go:318]: Yeah.. this was a scary scenario indeed where a boltdb transaction could have rolled back but the cache was updated anyway.

File: beacon-chain/db/kv/state_summary.go (hunk 3 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/state_summary.go:57]: sounds good, thanks!

File: beacon-chain/blockchain/error.go [hunk 7]
```diff
@@ -23,6 +23,8 @@ var (
 	errNotOptimisticCandidate = errors.New("block is not suitable for optimistic sync")
 	// errBlockNotFoundInCacheOrDB is returned when a block is not found in the cache or DB.
 	errBlockNotFoundInCacheOrDB = errors.New("block not found in cache or db")
+	// errNilBlock is returned when a nil block is returned from the cache.
```
Comment (potuz) [inline beacon-chain/blockchain/error.go:26]:
```suggestion
	// errNilBlockInCache is returned when a nil block is returned from the cache.
```

File: beacon-chain/db/kv/blocks.go [hunk 8]
```diff
@@ -296,32 +296,40 @@ func (s *Store) SaveBlocks(ctx context.Context, blks []interfaces.SignedBeaconBl
...
+			blindedBlock, err := blk.ToBlinded()
+			if err != nil {
+				if !errors.Is(err, blocks.ErrUnsupportedVersion) {
+					return err
+				}
+			} else {
+				blk = blindedBlock
+			}
+		}
+		s.blockCache.Set(string(blockRoots[i]), blk, 0)
```
Comment (potuz) [inline beacon-chain/db/kv/blocks.go:330]: Why the change in the cost from the previous behavior of `int64(len(encodedBlocks[i]))`

File: beacon-chain/db/kv/state_summary.go [hunk 9]
```diff
@@ -21,10 +25,20 @@ func (s *Store) SaveStateSummary(ctx context.Context, summary *ethpb.StateSummar
...
+
+// SaveStateSummariesWithPendingBlocks saves state summary objects to the DB.
+func (s *Store) SaveStateSummariesWithPendingBlocks(ctx context.Context, summaries []*ethpb.StateSummary, blockCache func([32]byte) (interfaces.SignedBeaconBlock, error)) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.SaveStateSummariesWithPendingBlocks")
+	defer span.End()
 
 	// When we reach the state summary cache prune count,
 	// dump the cached state summaries to the DB.
 	if s.stateSummaryCache.len() >= stateSummaryCachePruneCount {
+		if err := s.ensureBlocksSaved(ctx, blockCache); err != nil {
```
Comment (potuz) [inline beacon-chain/db/kv/state_summary.go:39]: This function seems to ignore the blocks that are not in the cache and are not in the summaries. It will not return an error and the summaries will be saved down below without a corresponding block in DB

File: beacon-chain/db/kv/blocks.go (hunk 8 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/blocks.go:330]: Now that I have slept since looking at this, I understand it. I will revert that cache costing change. It didn't make any sense to me last night but I get it now

File: beacon-chain/db/kv/state_summary.go (hunk 9 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/state_summary.go:39]: See line 137 for the check that blocks are in the db.

File: beacon-chain/db/kv/blocks.go (hunk 8 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/blocks.go]: Wait, I looked at it again and I dont understand the costing now.

Why is 1 block set to the cost of `int64(len(encodedBlocks[i]))`?

Comment (prestonvanloon) [reply beacon-chain/db/kv/blocks.go:330]: oh wait, because it's the length of the encoded block... i see now

**Review by terencechain — COMMENTED** [review]
This looks largely good to me. Might worth picking it up again

**Comment by potuz** [comment]
> This looks largely good to me. Might worth picking it up again

CC @kasey will this conflict with your work?
